From: Ilya Leoshkevich <iii@linux.ibm.com>
Subject: s390/diag: unpoison diag224() output buffer
Date: Thu, 14 Dec 2023 00:24:45 +0100

Diagnose 224 stores 4k bytes, which cannot be deduced from the inline
assembly constraints. This leads to KMSAN false positives.

Unpoison the output buffer manually with kmsan_unpoison_memory().

Link: https://lkml.kernel.org/r/20231213233605.661251-26-iii@linux.ibm.com
Signed-off-by: Ilya Leoshkevich <iii@linux.ibm.com>
Cc: Alexander Gordeev <agordeev@linux.ibm.com>
Cc: Alexander Potapenko <glider@google.com>
Cc: Christian Borntraeger <borntraeger@linux.ibm.com>
Cc: Christoph Lameter <cl@linux.com>
Cc: David Rientjes <rientjes@google.com>
Cc: Dmitry Vyukov <dvyukov@google.com>
Cc: Heiko Carstens <hca@linux.ibm.com>
Cc: Hyeonggon Yoo <42.hyeyoo@gmail.com>
Cc: Joonsoo Kim <iamjoonsoo.kim@lge.com>
Cc: Marco Elver <elver@google.com>
Cc: Mark Rutland <mark.rutland@arm.com>
Cc: Masami Hiramatsu <mhiramat@kernel.org>
Cc: Pekka Enberg <penberg@kernel.org>
Cc: Roman Gushchin <roman.gushchin@linux.dev>
Cc: Steven Rostedt (Google) <rostedt@goodmis.org>
Cc: Sven Schnelle <svens@linux.ibm.com>
Cc: Vasily Gorbik <gor@linux.ibm.com>
Cc: Vlastimil Babka <vbabka@suse.cz>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 arch/s390/kernel/diag.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/arch/s390/kernel/diag.c~s390-diag-unpoison-diag224-output-buffer
+++ a/arch/s390/kernel/diag.c
@@ -9,6 +9,7 @@
 #include <linux/export.h>
 #include <linux/init.h>
 #include <linux/cpu.h>
+#include <linux/kmsan-checks.h>
 #include <linux/seq_file.h>
 #include <linux/debugfs.h>
 #include <linux/vmalloc.h>
@@ -255,6 +256,7 @@ int diag224(void *ptr)
 		"1:\n"
 		EX_TABLE(0b,1b)
 		: "+d" (rc) :"d" (0), "d" (addr) : "memory");
+	kmsan_unpoison_memory(ptr, PAGE_SIZE);
 	return rc;
 }
 EXPORT_SYMBOL(diag224);
_
